modules/edu.gemini.seqexec.web/edu.gemini.seqexec.web.server/src/main/scala/edu/gemini/seqexec/web/server/security/FreeLDAPAuthenticationService.scala

```diff
@@ -0,0 +1,119 @@
+package edu.gemini.seqexec.web.server.security
+
+import com.unboundid.ldap.sdk._
+import edu.gemini.seqexec.web.common.UserDetails
+import edu.gemini.seqexec.web.server.security.AuthenticationService.AuthResult
+
+import scalaz.Scalaz._
+import scalaz._
+import scalaz.concurrent.Task
+import scalaz.effect.IO
+
+/**
+  * Definition of LDAP as a free monad algebra/interpreters
+  */
+object FreeLDAPAuthenticationService {
+  import UserDetails._
+  import LdapConnectionOps._
+
+  sealed trait LdapOp[A]
+  // Operations on ldap
+  object LdapOp {
+    // Operation to authenticate a user, It returns true if it works
+    case class AuthenticateOp(username: String, password: String) extends LdapOp[UID]
+    // Read the user display name
+    case class UserDisplayNameOp(uid: UID)                        extends LdapOp[DisplayName]
+    // Reads the name, groups and thumbnail
+    case class DisplayNameGrpThumbOp(uid: UID)                    extends LdapOp[(DisplayName, Groups, Option[Thumbnail])]
+  }
+
+  // Free monad over the free functor of LdapOp.
+  type LdapM[A] = Free[LdapOp, A]
+
+  // Smart constructors for LdapOp[A]
+  def bind(u: String, p: String): LdapM[UID]                                   = Free.liftF(LdapOp.AuthenticateOp(u, p))
+  def displayName(u: UID): LdapM[DisplayName]                                  = Free.liftF(LdapOp.UserDisplayNameOp(u))
+  def nameGroupsThumb(u: UID): LdapM[(DisplayName, Groups, Option[Thumbnail])] = Free.liftF(LdapOp.DisplayNameGrpThumbOp(u))
+
+  // Natural transformation to IO
+  def toIO(c: LDAPConnection): LdapOp ~> IO =
+    new (LdapOp ~> IO) {
+      def apply[A](fa: LdapOp[A]) =
+        fa match {
+          case LdapOp.AuthenticateOp(u, p)       => IO(c.authenticate(u, p))
+          case LdapOp.UserDisplayNameOp(uid)     => IO(c.displayName(uid))
+          case LdapOp.DisplayNameGrpThumbOp(uid) => IO(c.nameGroupsThumb(uid))
+      }
+    }
+
+  // Natural transformation to Task
+  def toTask(c: LDAPConnection): LdapOp ~> Task =
+    new (LdapOp ~> Task) {
+      def apply[A](fa: LdapOp[A]) =
+        fa match {
+          case LdapOp.AuthenticateOp(u, p)       => Task(c.authenticate(u, p))
```

Comment:
You should use `Task.delay` here; `Task.apply` will cause the operation to run on an implicit thread pool which probably isn't what you want. In fact the whole interpreter could be `Task.delay(toIO(c).unsafePerformIO)` if you're feeling lazy.